crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,248 @@
+                    Some(*query)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        for query in &completed_and_ignored_requests {
+            active_requests.remove(query);
+        }
+    }
+}
+
+/// A price estimator which doesn't issue another estimation request while an identical one is
+/// already in-flight.
+pub struct BufferingPriceEstimator {
+    inner: Arc<Inner>,
+}
+
+impl BufferingPriceEstimator {
+    pub fn new(estimator: Box<dyn PriceEstimating>) -> Self {
+        Self {
+            inner: Arc::new(Inner {
+                estimator,
+                in_flight_requests: Mutex::new(Default::default()),
+            }),
+        }
+    }
+
+    async fn estimate_buffered(&self, queries: &[Query]) -> Vec<EstimationResult> {
+        // For each `Query` either get an in-flight request or keep the `Query` to forward it to the
+        // inner price estimator.
+        let (active_requests, remaining_queries): (Vec<_>, Vec<_>) = {
+            let requests = self.inner.in_flight_requests.lock().unwrap();
+            queries
+                .iter()
+                .map(|query| match requests.get(query).cloned() {
+                    Some(active_request) => (Some(active_request), None),
+                    None => (None, Some(*query)),
+                })
+                .unzip()
+        };
+
+        // Create future which estimates all `remaining_queries` in a single batch.
+        let fetch_remaining_estimates = {
+            let remaining_queries: Vec<_> = remaining_queries.iter().flatten().cloned().collect();
+            let inner = self.inner.clone();
+            async move { inner.estimator.estimates(&remaining_queries).await }
+                .boxed()
+                .shared()
+        };
+
+        // Create a `SharedEstimationRequest` for each individual `Query` of the batch. This
+        // makes it possible for a `batch_2` to await the queries which it is interested in of the
+        // in-flight `batch_1`. Even if the estimator which requested `batch_1` stops polling it, the
+        // estimator of `batch_2` can still poll `batch_1` to completion by polling the
+        // `SharedEstimationRequest` it is actually interested in.
+        //
+        // Build those shared futures up front to keep the critical section short when inserting
+        // them into the `active_requests`.
+        #[allow(clippy::needless_collect)]
+        let individual_requests_for_batch: Vec<_> = remaining_queries
+            .iter()
+            .flatten()
+            .enumerate()
+            .map(|(index, query)| {
+                let fetch_remaining_estimates = fetch_remaining_estimates.clone();
+                (
+                    *query,
+                    async move { fetch_remaining_estimates.await[index].clone() }
+                        .boxed()
+                        .shared(),
```

Comment:
I did a double take on the double boxing/sharing of the future, but I think I understand why it is needed: We are creating a single batched query future and so it needs to be shared for each query of the batch (which also needs to be shared for the `inflight_requests_cache`.